### PR TITLE
Refactor the key event handling

### PR DESCRIPTION
**Pull request by bbrto21**
* Removes platform dependencies from all SendKeyEvent methods.
* Adds methods related to composing a text to the FlutterView and
  TextInputChannel.
* Transfers the key event filter role to TizenInputMethodContext.
* Adds filter methods for each event type to TizenInputMethodContext.
* Removes the concept of selection mode for text input.

### CODE REVIEW

**Comment by bbrto21** [comment]
contribute to https://github.com/flutter-tizen/flutter-tizen/issues/351

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (JSUYA) [inline shell/platform/tizen/channels/text_input_channel.cc:75]: Prefer if (value) over if (value != nullptr). [link](https://github.com/flutter-tizen/plugins/issues/354)
So. how about `if (!active_model)`?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (JSUYA) [inline shell/platform/tizen/channels/text_input_channel.cc:74]: cursor_pos is not used in this method (also higher level methods). Is there any reason to keep it?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (JSUYA) [inline shell/platform/tizen/channels/text_input_channel.cc:78]: When I checked several cases of checking a string, it seems to use `empty()` to check an empty `string`.
Could you check it out?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (JSUYA) [inline shell/platform/tizen/channels/text_input_channel.cc:388]: model_ argument is not used in this method. do you need it?

File: shell/platform/tizen/tizen_input_method_context.cc
Comment (JSUYA) [inline shell/platform/tizen/tizen_input_method_context.cc:148]: please change `0` to `false`

File: shell/platform/tizen/tizen_window_elementary.cc
Comment (JSUYA) [inline shell/platform/tizen/tizen_window_elementary.cc:22]: `Contorol` looks like typo.

File: shell/platform/tizen/tizen_input_method_context.cc
Comment (JSUYA) [reply shell/platform/tizen/tizen_input_method_context.cc:148]: oh my mistake. please ignore this comment

File: shell/platform/tizen/flutter_tizen_view.cc
Comment (swift-kim) [inline shell/platform/tizen/flutter_tizen_view.cc:220]: These `SendKeyEvent` methods defined by each "channel" class have similar method signatures. How about defining a common interface for them?

It might be hard to define such interface because `TextInputChannel`'s `SendKeyEvent` is synchronous while `KeyEventChannel`'s `SendKeyEvent` is asynchronous.

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:75]: I always prefer `if (value)` than `if (value != nullptr)`.
but in this case, I prefer `if (value == nullptr)` than `if (!value)`, because it is used consistently, at least in this file, and the meaning is conveyed more clearly in situations where error handling is required.

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:74]: Pass the result of 'ecore_imf_context_preedit_string_get' as is. that's all. should I remove in this PR?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:78]: They have the same result.

File: shell/platform/tizen/public/flutter_platform_view.h
Comment (bbrto21) [reply shell/platform/tizen/public/flutter_platform_view.h:52]: I know. Should it be open at the same time as this PR?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:365]: vscode formatter is interfering with this, do you have a guide?

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:128]: If you don't mind, I'd like to deal with this as a different topic.

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:115]: It seems to be a legacy that was created when `TextInputChannel` was more complicated than necessary.
at that time, if you press the hardware back key to hide the keyboard, I remember that some clean up was needed. 
but I don't think it's all necessary now. I'll remove everything that's relevant.

File: shell/platform/tizen/channels/key_event_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/key_event_channel.cc:261]: Yes, but I decided to use the name used in the EFL. Should I change it?

File: shell/platform/tizen/flutter_tizen_view.cc
Comment (bbrto21) [reply shell/platform/tizen/flutter_tizen_view.cc:220]: I already thought about this, but I didn't include it in this PR. because this topic got me thinking too much :(
In the end, I excluded this in order to achieve the purpose of this PR quickly. But it's still on the list I want to refactor.

File: shell/platform/tizen/tizen_window_elementary.h
Comment (bbrto21) [reply shell/platform/tizen/tizen_window_elementary.h:66]: I haven't decided on this yet. because I was considering #292, I was planning to pick a suitable location for this after #291 and #292 merged. 

Can I do that? or should I add `TizenWindow` implementation file now?

File: shell/platform/tizen/tizen_input_method_context.h
Comment (bbrto21) [reply shell/platform/tizen/tizen_input_method_context.h]: I picked `ShouldIgnoreKey`

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc:343]: Text input in languages ​​other than English that requires composing by IME, such as Korean or Japanese, is already supported.

This part you mentioned is inevitably added logic because the behavior of 'IME' is different in the profile except for TV.
(IME does not filter the key events corresponding to the ASCII in Tizen except TV profile.)

I will contact you via messenger for additional explanations.

File: shell/platform/tizen/channels/platform_view_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/platform_view_channel.cc:91]: I would like to consider this topic along with [this](https://github.com/flutter-tizen/engine/pull/291#discussion_r884389195).

File: shell/platform/tizen/tizen_window.h
Comment (JSUYA) [inline shell/platform/tizen/tizen_window.h:68]: How about `GetInputMethodConetext()` or `GetInputMethodConetextPtr()`?

Comment (bbrto21) [reply shell/platform/tizen/tizen_window.h:68]: we don't use xxPtr postfix in method name. we name get/setter according to [this](https://google.github.io/styleguide/cppguide.html#Function_Names) 

In particular, for getters that only return references from unique pointers, we are using this format widely.

File: shell/platform/tizen/channels/platform_view_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/platform_view_channel.cc:91]: Ok.
Changing the name of the `DispatchKeyDownEvent` was not a consideration for me.

This interface was created to support `webview` implementation in Tizen, and because these names are derived from the names of the widely known webview interface and their internal methods.

public boolean dispatchKeyEvent ([KeyEvent](https://developer.android.com/reference/android/view/KeyEvent) event)

Furthermore, I've never seen a place where 'Hook' is used except in the Windows world.

File: shell/platform/tizen/tizen_input_method_context.cc
Comment (JSUYA) [inline shell/platform/tizen/tizen_input_method_context.cc:106]: This `window_` is undefined.

File: shell/platform/tizen/tizen_window.h
Comment (JSUYA) [reply shell/platform/tizen/tizen_window.h:68]: oh.. ok,,

File: shell/platform/tizen/tizen_input_method_context.cc
Comment (bbrto21) [reply shell/platform/tizen/tizen_input_method_context.cc:106]: Thanks!

File: shell/platform/tizen/tizen_input_method_context.cc
Comment (bbrto21) [reply shell/platform/tizen/tizen_input_method_context.cc:150]: I added guard for wearable profile.

File: shell/platform/tizen/channels/platform_view_channel.cc
Comment (bwikbs) [reply shell/platform/tizen/channels/platform_view_channel.cc:91]: IMO, ``DispatchKeyDownEvent`` is not so bad. 
At the same time, I think @swift-kim 's suggestion can be accepted.
( ``KeyboardHook`` is a bit less familiar... to me)

**Review by bwikbs — APPROVED** [review]
There are minor disagreements at here 😄  
(In my case) It's pretty good to me.
Thanks for your hard working!! 👍

File: shell/platform/tizen/channels/platform_view_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/platform_view_channel.cc]: I suggest using `SendKey' as a name universally except for the callback function.

File: shell/platform/tizen/tizen_window_elementary.h
Comment (swift-kim) [reply shell/platform/tizen/tizen_window_elementary.h:66]: Okay, no problem!

File: shell/platform/tizen/public/flutter_platform_view.h
Comment (swift-kim) [reply shell/platform/tizen/public/flutter_platform_view.h]: Yes, it's recommended. At least the PRs should be ready before updating the tool's engine hash.

**Review by JSUYA — APPROVED** [review]
I have checked the modifications to the codes I reviewed.
looks good to me
wait for swift-kim 's approve

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (swift-kim) [reply shell/platform/tizen/channels/text_input_channel.cc]: Hmm, really? There's no problem in my case and nothing is detected by `ci/format.sh` either.

image

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (swift-kim) [reply shell/platform/tizen/channels/text_input_channel.cc:128]: Okay, that's great.

File: shell/platform/tizen/channels/key_event_channel.cc
Comment (swift-kim) [reply shell/platform/tizen/channels/key_event_channel.cc]: I'd recommend so. At least we can avoid a naming conflict in this method (`keycode` vs. `key_code`).

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/text_input_channel.cc]: done

File: shell/platform/tizen/channels/key_event_channel.cc
Comment (bbrto21) [reply shell/platform/tizen/channels/key_event_channel.cc]: please check new changes.

File: shell/platform/tizen/tizen_input_method_context.h
Comment (swift-kim) [reply shell/platform/tizen/tizen_input_method_context.h]: Do you want to keep the name `HandleUnfilteredEvent` as it is because that part will be reworked soon? https://github.com/flutter-tizen/engine/pull/291#discussion_r883870830

Comment (bbrto21) [reply shell/platform/tizen/tizen_input_method_context.h]: Thank you for the detailed review, I changed it. please review again.

File: shell/platform/tizen/public/flutter_platform_view.h
Comment (bbrto21) [reply shell/platform/tizen/public/flutter_platform_view.h]: I'm working on this. I'll create PR as soon as possible.

File: shell/platform/tizen/channels/text_input_channel.cc
Comment (swift-kim) [inline shell/platform/tizen/channels/text_input_channel.cc:305]: Here's a typo.
```suggestion
                                 uint32_t modifiers) {
```

File: shell/platform/tizen/tizen_input_method_context.h
Comment (swift-kim) [reply shell/platform/tizen/tizen_input_method_context.h]: Okay, thanks!

File: .github/workflows/build.yml
Comment (bbrto21) [inline .github/workflows/build.yml:50]: I removed the cache action because the build jobs on CI repeatedly failed.

CC. @WonyoungChoi

Comment (bbrto21) [reply .github/workflows/build.yml:50]: @WonyoungChoi informed me that this issue is being tracked [here](https://github.com/actions/cache/issues/811).

and he and me agreed to temporarily remove the cache action.

**Review by swift-kim — APPROVED** [review]

**Pull request merged**
